docs/new_protocol.asciidoc

```diff
@@ -0,0 +1,463 @@
+	ReceivedFin(tcptuple *common.TcpTuple, dir uint8,
+		private ProtocolData) ProtocolData
+
+	// Called when a packets are missing from the tcp
+	// stream.
+	GapInStream(tcptuple *common.TcpTuple, dir uint8, nbytes int,
+		private ProtocolData) (priv ProtocolData, drop bool)
+}
+
+type UdpProtocolPlugin interface {
+	ProtocolPlugin
+
+	// ParseUdp is invoked when UDP payload data is available for parsing.
+	ParseUdp(pkt *Packet)
+}
+----------------------------------------------------------------------
+
+At the high level, the protocols plugins receive raw packet data via the
+`Parse()` or `ParseUdp()` methods and produce objects that will be indexed into
+Elasticsearch via the `results` channel.
+
+The `Parse()` and `ParseUdp()` methods are called for every packet that is
+sniffed and is using one of the ports of the protocol as defined by the
+`GetPorts()` function. They receive the packet in a Packet object, which looks
+like this:
+
+[source,go]
+----------------------------------------------------------------------
+type Packet struct {
+	Ts      time.Time <1>
+	Tuple   common.IpPortTuple <2>
+	Payload []byte <3>
+}
+----------------------------------------------------------------------
+
+<1> The timestamp of the packet
+<2> The source IP address, source port, destination IP address, destination port
+combination.
+<3> The application layer payload (i.e. without the IP and TCP/UDP headers) as a
+byte slice.
+
+The objects sent through the `results` channel have the type `common.MapStr`
+which is essentially a `map[string]interface{}` with a few more convenience
+https://github.com/elastic/libbeat/blob/fae9cf861b58f09cf578245e45415899f4151d32/common/mapstr.go[methods]
+added.
+
+Besides the `Parse()` function, the TCP layer also calls the `ReceivedFin()`
+when a TCP stream is closed and the `GapInStream()` functions when packet loss
+is detected in a TCP stream. The protocol module can use these callbacks to take
+decisions about what to do with partial data received. For example, for the
+HTTP/1.0 protocol, the end of connection is used to know when the message is
+finished.
+
+
+==== The TCP Parse function
+
+For TCP protocols, the `Parse()` function is the heart of the module. As
+mentioned before, it is called for every TCP packet containing data on the
+configured ports.
+
+It is important to understand that because TCP is a stream
+based protocol, the packets boundaries don't necessarily match the application
+layer message boundaries. For example, a packet can contain only a part of the
+message, it can contain a complete message or it can contain multiple messages.
+
+If you see a packet in the middle of the stream, you have no guaranties that its
+first byte is the beginning of a message. However, if the packet is the first
+seen in a given TCP stream, that you can assume is the beginning of the message.
+
+The Parse function needs to deal with these facts which generally means that it
+needs to keep state across multiple packets.
```

Comment:
see applayer and streambuf modules (still in memcache development branches) for some support in keeping state.

Also make sure critical parts in parser state are somehow introspectible (depending on parsing strategy) to handle Fin and stream gaps